### PR TITLE
Merge decisions

### PR DESCRIPTION
**Pull request by vidartf**
Implements merging using the merge decision format described in #74 (see also #75).

### Todo:
 - [x] Have autoresolve work with the new format.
 - [x] Update methods/tests that want both conflicts and merged output to use the new interface.
 - [x] Ensure parity with old method.
 - [x] Make new test / update old ones to test decision making and application separately.

### CODE REVIEW

**Comment by vidartf** [comment]
This PR should be ready for review/merge now.

File: nbdime/diff_format.py [hunk 1]
```diff
@@ -293,57 +293,76 @@ def _accum(seq):
...
+
 
 def _combine_ops(existing, new):
     """Combines new op into existing op
     """
-    if new.op == DiffOp.ADDRANGE:
-        existing.valuelist += new.valuelist
+    if new.op in _addops:
+        if existing.op == DiffOp.ADD:
+            existing.op == DiffOp.ADDRANGE
```
Comment (martinal) [inline nbdime/diff_format.py]: This looks like a bug.

File: nbdime/diff_format.py [hunk 2]
```diff
@@ -293,57 +293,76 @@ def _accum(seq):
...
+            existing.valuelist = [existing.value]
+            del existing.value
```
Comment (martinal) [inline nbdime/diff_format.py]: `del` won't work on `DiffEntry`, I'd rewrite the code to consider existing immutable and recreate a new diffentry object instead.

File: nbdime/diff_format.py [hunk 3]
```diff
@@ -293,57 +293,76 @@ def _accum(seq):
...
     """Translates a diff of strings split by str.splitlines() to a diff of
     the joined multiline string
     """
+    if isinstance(a, string_types):
+        a = a.splitlines(True)
     a_mapping = [0] + list(_accum(len(ia) for ia in a))
     flattened = []
     for e in diff:
         op = e.op
         new_key = a_mapping[e.key]
```
Comment (martinal) [inline nbdime/diff_format.py]: Can I suggest renaming `a_mapping` to e.g. `line_to_char` and `new_key` to e.g. `char_key`?

File: nbdime/diff_format.py [hunk 4]
```diff
@@ -293,57 +293,76 @@ def _accum(seq):
...
+                oo = _check_overlaps(flattened, d)
+                if oo is None:
+                    flattened.append(d)
+                else:
+                    _combine_ops(oo, d)
+        else:
             d = copy.deepcopy(e)
             d.key = new_key
             if op == DiffOp.ADDRANGE:
                 d.valuelist = "".join(e.valuelist)
```
Comment (martinal) [inline nbdime/diff_format.py]: I'd prefer a non-destructive style instead:
`d = op_addrange(new_key, "".join(e.valuelist))`

File: nbdime/diff_format.py [hunk 5]
```diff
@@ -293,57 +293,76 @@ def _accum(seq):
...
-            else:
+            elif op == DiffOp.REMOVERANGE:
                 d.length = a_mapping[e.key + e.length] - d.key
```
Comment (martinal) [inline nbdime/diff_format.py]: Same: `d = op_removerange(new_key, a_mapping[e.key + e.length] - d.key)`

File: nbdime/diff_format.py [hunk 6]
```diff
@@ -293,57 +293,76 @@ def _accum(seq):
...
+        if new.op == DiffOp.ADDRANGE:
+            existing.valuelist += new.valuelist
+        else:
+            if isinstance(existing.valuelist, string_types):
+                existing.valuelist += new.value
+            else:
+                existing.valuelist.append(new.value)
     elif new.op == DiffOp.REMOVERANGE:
-        existing.lenght += new.length
+        existing.length += new.length
```
Comment (martinal) [inline nbdime/diff_format.py]: Are you sure this is right? If I follow the code correctly, seems to me that here `new.key == existing.key` as checked in `_check_overlaps`, so adding the lengths here will extend the removerange operation to chars that were previously not deleted?

**Comment by vidartf** [comment]
@martinal I think this should be ready for further review now. I tried to update the docstrings as well, to make review easier. Specific tests of the functional parts should probably be expanded as well, but for now they are tested through overall functionality.

File: nbdime/diffing/generic.py [hunk 7]
```diff
@@ -78,7 +78,7 @@ def diff(a, b, path="", predicates=None, differs=None):
     elif isinstance(a, string_types) and isinstance(b, string_types):
         # FIXME: Do we need this string case, and if so do we need to pass on
```
Comment (martinal) [inline nbdime/diffing/generic.py]: I guess this fixme can be removed now.

File: nbdime/merging/autoresolve.py [hunk 8]
```diff
@@ -123,148 +143,160 @@ def make_inline_source_value(base, le, re):  # FIXME: Test this!
...
+        dec.action = "local_then_remote"
+        dec.conflict = False
+    elif strategy == "clear":
+        dec.action = "clear"
+        dec.conflict = False
+    elif strategy == "mergetool":
+        # Leave this type of conflict for external tool to resolve
+        pass
+    elif strategy == "clear-parent":
+        dec.action = "clear_parent"
```
Comment (martinal) [inline nbdime/merging/autoresolve.py]: Should we use either - or _ consistently?

File: nbdime/merging/autoresolve.py [hunk 9]
```diff
@@ -123,148 +143,160 @@ def make_inline_source_value(base, le, re):  # FIXME: Test this!
...
+        dec.conflict = False
     elif strategy == "fail":
-        raise RuntimeError("Not expecting a conflict at path {}.".format(path))
+        raise RuntimeError("Not expecting a conflict at path {}.".format(dec.common_path + '/*'))
```
Comment (martinal) [inline nbdime/merging/autoresolve.py]: `dec.common_path` seems to be a tuple in other code?

File: nbdime/merging/decisions.py [hunk 10]
```diff
@@ -0,0 +1,462 @@
...
+    decision level.
+
+    If the decision is already at the lowest level, it returns the original
+    decision.
+    """
+    try:
+        while 1:
+            decision = pop_patch_decision(decision)
+    except ValueError as e:
+        if not str(e).startswith("Cannot pop patch decision for: "):
```
Comment (martinal) [inline nbdime/merging/decisions.py]: I think you already rewrote this in the other branch, but a sidenote on exception usage: when you want to catch a particular exception, rather make your own exception subclass and catch that type, e.g.:
```
class MyFailure(Exception):
    pass

def f():
    raise MyFailure(message)

try:
    f()
except MyFailure as e:
    handle exception
```

**Pull request merged**